tracer/publisher.go

```diff
@@ -0,0 +1,112 @@
+package tracer
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"math/rand"
+	"net/http"
+	"os"
+	"sync/atomic"
+	"time"
+
+	"nhooyr.io/websocket"
+	"subtrace.dev/pubsub"
+	"subtrace.dev/rpc"
+)
+
+var DefaultPublisher = newPublisher()
+
+type Publisher struct {
+	log atomic.Bool
+	ws  *websocket.Conn
+
+	ch chan *[]byte
+}
+
+func newPublisher() *Publisher {
+	p := &Publisher{ch: make(chan *[]byte)}
```

Comment:
the channel needs to be buffered so that we don't accumulate a goroutine for every request the server handles